Pass vlan tagged packets back to the kernel stack.

kvaster opened the pull request:
VLAN packets will be catched with bpf on main interface first. We
need to passs it back to the kernel stack. VLAN info/tag will be stripped and
packet will be reenqueued on proper interface or dropped.

Also we do not need to process VLAN packets on egress, cause either
we've already processed such packets in case bpf program is attached
to VLAN interface or we do not need to process such packets at all in
other way.

VLAN tagged packets will be processed in the folowing way:
* If vlan tag is zero packet will be processed as usual (cause it is targeted for main interface)
* For each device will be allowed only corresponding vlan tags.
* By default will be allowed only tags from devices controlled by cilium.
* Any other tags may be allowed via option (`--vlan-bpf-bypass`), `--vlan-bpf-bypass 0` may be used for allowing all available vlan tags.
* All other vlan tags will be always dropped for security reason.

Concerns:
* Probably ifindexes and vlan tags should be sorted to make configuration predictable.
* Probably something like `*` should be used insteadof `0` for allowing all available vlan tags.

Previous pull request and discussion: https://github.com/cilium/cilium/pull/15534

Fixes: cilium#14579

Signed-off-by: Viktor Kuzmin <kvaster@gmail.com>

```release-note
Make NodePort BPF to work on VLAN devices
```

twpayne commented on pkg/datapath/linux/config/config.go:
```diff
@@ -549,6 +555,63 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+	}
+
+	if len(vlansByIfIndex) == 0 {
+		return "return CTX_ACT_DROP;", nil
```
The description of the `--bpf-allow-vlan` flag is "List of explicitly allowed VLAN IDs, zero will allow all VLAN IDs", but won't the `CTX_ACT_DROP` here mean that zero will allow *no* VLAN IDs? i.e. should this be a `CTX_ACT_OK` instead?

kvaster replied:
Bad description. I will correct - `--bpf-allow-vlan 0` means all VLAN IDs.

kvaster replied:
And some notices: `len(vlansByIfIndex) == 0` is true only when there is no any vlan device corresponding to main devices. By default all vlan devices are added if they are controlled by cilium. But there might be case when there will be vlan devices wich are not controlled by cilium - in this case we can pass option - `--bpf-allow-vlan 4000,4001` or allow all vlan devices - `--bpf-allow-vlan 0`

twpayne reviewed (approved):
Looks good from a CLI perspective, but I need the experienced datapath folk to review the datapath aspects.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -549,6 +555,63 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
+// vlanFilterMacros generates VLAN_FILTER macros which
+// are written to node_config.h
+func vlanFilterMacros() (string, error) {
```
Is there a way to unit test this logic?

joestringer replied:
One path for this would be, do something similar to existing privileged unit tests, involving creating dummy devices and so on so that the code can actually perform queries against a configured Linux system, and then generate appropriate output. Then make sure that it picks up the appropriate devices depending on whether we're controlling the main device or not.

Separately there's also the question of whether the output of this function is valid C, ideally we could either lint or even compile a snippet as part of automated testing. Maybe at this point we're not really talking unit tests but more of an integration test.

In future maybe we could even support dynamically configured VLANs and we could test that support.

kvaster replied:
Unfortunatelly I have no good experiance with cilium yet and I think it will take much time for me to write such tests... The output of this function is valid C.

joestringer replied:
Would you be up for trying to do at least some minimal tests like configure a specific set of VLAN devices, then run this logic, then check that the output of the function is detecting the correct devices?

There's some fairly basic "create devices" type logic in `pkg/datapath/link/link_test.go`, you would need to go a bit further to create the appropriate VLAN devices. The docs have some examples about how you can run those individual privileged tests to validate the behaviour locally[0].

[0] https://docs.cilium.io/en/latest/contributing/testing/unit/#testing-individual-packages

kvaster replied:
Ok. Thanks. I will do.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -980,6 +980,9 @@ func init() {
 	flags.Bool(option.ExternalClusterIPName, false, "Enable external access to ClusterIP services (default false)")
 	option.BindEnv(option.ExternalClusterIPName)
 
+	flags.IntSlice(option.BPFAllowVLAN, []int{}, "List of explicitly allowed VLAN IDs, '0' id will allow all VLAN IDs")
```
Could we just treat the empty list as allow all?

So basically if you don't configure the option, the behaviour is exactly the same as today: No filtering at all. If you configure the option, you are writing an explicit list of VLANs to allow.

kvaster replied:
I think, no, according to talks on previous pull request. Today's behaviour doesn't mean 'allow all'. Today's behaviour is broken - cilium treats any vlan tagged packets as packets without tag and may take packet from main interface and resend to another node (for example) from vlan tagged interface and this will add one more vlan tag -> packet processing will be broken. Current behaviour does not work at all with vlan tagged packets.

joestringer replied:
OK, in that case if it's pretty broken today then it'd be better to drop early here with clear drop notifications so if someone configures this way, they at least get a more obvious failure they can investigate rather than just weird packets turning up in unexpected places. :+1:

joestringer commented on pkg/option/config.go:
```diff
@@ -950,6 +950,9 @@ const (
 	ExternalClusterIPName = "bpf-lb-external-clusterip"
+
+	// BPFAllowVLAN instructs Cilium to allow vlan targeted packets
+	BPFAllowVLAN = "bpf-allow-vlan"
```
I think we need a clearer name and semantics on this. Until I got to the review of the actual code being generated, I was under the impression that this functionality was a filter in the policy sense: Ie, decide when to drop things, using an inclusion list.

In practice what it's configuring is a bypass of Cilium logic. Maybe something like "vlan-bpf-bypass" or "vlan-bpf-skip"?

kvaster replied:
I was thinking about `vlan-bpf-bypass` first, but current logic will malform vlan tagged packets and we need to drop or to bypass them to kernel. Think I need help in this question :)

joestringer replied:
I think that technically there's two changes occurring in this PR:
a) Resolve the "malformed VLAN-tagged packets issue" by dropping packets with VLAN tags.
b) Provide a configuration option to bypass Cilium BPF logic.

(a) is basically a new default and I reason above[[link][0]] that this will just make it easier to detect when VLAN misconfiguration is causing issues in your network. (b) is then saying, if you know how the Linux stack is configured to handle VLAN traffic and you're OK with the implications, then you can teach Cilium to bypass its BPF logic for VLAN traffic.

I feel like longer term there is probably a more cohesive solution that involves teaching Cilium more about the overall VLAN configuration in your network and reasoning about the overall system + Cilium's part in it. But for now this combination of (a) / (b) should be enough to unblock what you're trying to do. And I'd argue that (b) is really what the configuration option is about.

[0]: https://github.com/cilium/cilium/pull/16772#discussion_r668158630

kvaster replied:
Hm. It should be a good 'fix' for those who require vlan's. I will remove this complex logic and will make bpf drop or bypass all vlan tags depending on option with default to drop. Thanks for comments, will get back soon.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -997,8 +1012,18 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+	if (ctx->vlan_present) {
+		vlan_id = ctx->vlan_tci & 0xfff;
+		if (vlan_id) {
+			VLAN_FILTER(ctx->ifindex, vlan_id);
```
I'd like to make sure we don't break existing behaviour here. It looks to me like by default if the user has not configured the option, we will now start dropping all vlan-tagged traffic. Maybe we need to differentiate between whether the option is enabled (which enables default drop for vlan-tagged traffic) vs. not configured (which should have no behaviour change)?

kvaster replied:
As I wrote earlier, keeping existing behavious make no sense, cause current behaviour makes malformed packets with vlan tags involved.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -968,6 +972,17 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
+	if (ctx->vlan_present) {
+		vlan_id = ctx->vlan_tci & 0xfff;
+		if (vlan_id) {
+			VLAN_FILTER(ctx->ifindex, vlan_id);
```
nit: As a matter of style, it's difficult to understand the implications of this `VLAN_FILTER` unless you know the details of how it works (which is autogenerated, so even harder to understand). Can we make this more explicit in the coding here?

```
ret = VLAN_FILTER(...)
if (ret < 0) {
 // .... drop path
} else if (ret > 0) {
  // ... bypass path
}
```

(Or switch case if you prefer, the point is not to hijack the execution logic via macro)

kvaster replied:
Will make changes!

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -549,6 +555,63 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+break; \
+} {{end}} \
+} \
+return CTX_ACT_DROP;`))
```
Naked drops like this will be hard to trace. Usually whenever we return a drop error code, we also `send_drop_notify(...)` with a dedicated drop code indicating the intent of _why_ we dropped this packet in this case. We probably need a new `DROP_VLAN_FILTERED` error code in `bpf/lib/common.h` and a good error message in `pkg/monitor/api/drop.go`, maybe something like "VLAN traffic disallowed by VLAN filter".

Could you update all of the `CTX_ACT_DROP` cases here to also send a drop notification?

joestringer replied:
(Note, usually this is handled by the outer code once you return an error code back up the stack... but with the current datapath implementation in this PR, this logic is invoked from the top level so there is no code which would catch the error and perform the send_drop_notify).

joestringer reviewed (changes requested):
This seems much closer to striking the right balance between providing the configurability you're looking for without surprising users too much with traffic bypassing Cilium logic.

I've requested a few changes below to make the change easier to reason about, both from a user perspective (semantics/naming and traceability) but also from a developer perspective (make it super simple to follow the flow of logic, test it so we can be sure that different variations of the option produce the desired behaviour).

The only thing I didn't raise below is the complexity concern. I strongly suspect that if you configure too many VLANs in this option, bpf program load will fail and the Cilium container will crash out because it cannot configure the datapath in the way you have requested. This could happen either when you first enable the option (because you specified too many tags), or it could plausibly happen later on when you upgrade to a new release, because each release can change the datapath. When you upgrade, there could be just enough additional BPF instructions introduced by this logic that when combined with the rest of the new datapath logic, it pushes the program complexity over some limit. I don't know where that limit is, and this PR is not proposing any testing for it. For now if this is a super-advanced option, maybe you're willing to take on that risk and deal with the consequences / roll back & report issues upstream when you hit them during upgrade. But we may want to think a bit more about integration testing and how we can detect such problems earlier in the development cycle so we can start the discussions early on how to address them.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -549,6 +555,63 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+
+	allowedVlans := make(map[int]bool)
+	for _, vlanId := range option.Config.BPFAllowVLAN {
+		allowedVlans[vlanId] = true
```
One way to make the "allow all" logic a bit nicer would be to just support "all" as one of the options for the flag. If "all" shows up in this list then rather than iterating the devices and generating a list of specific VLANs to pass through, we can generate `return CTX_ACT_OK` if any VLAN is set.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -37,6 +37,10 @@
 #define SKIP_ICMPV6_ECHO_HANDLING
 
+#ifndef VLAN_FILTER
+# define VLAN_FILTER(ifindex, vlan_id) return CTX_ACT_DROP;
```
You might want to consider adding a more complex default in the `bpf/node_config.h` file. That file is only used for testing, and is used in existing verifier tests. So if you define `VLAN_FILTER(...)` there with the most common configuration you would expect (could be 1-2 or maybe even 5 VLAN tags), then this would help us to catch potential verifier issues for common use cases before we release & you deploy them.

kvaster replied:
Common configuration from my point of view is one main device and one vlan device - internal and external networks. We faced that problem with our setup on hetzner datacenter. Telling the truth, I can't imaging configuration with lot's of vlan devices for kubenetes.

joestringer replied:
:+1: Given that VLANs have so much legacy usage, I suspect that in users own datacenters, additional use cases could come up with any number of VLAN configurations. I also note that by default this kind of configuration is configured once for the entire k8s cluster via cilium-config ConfigMap, and in the code here we configure it once for the entire node by generating the code into `node_config.h`. There's no sort of mapping between specific devices and specific VLAN tags. This means that the overall complexity is determined by the total number of VLAN tags that you expect your k8s cluster to encounter, and most likely every device on every node will have the same configuration.

If this is OK with you, then I see this as an acceptable limitation of the current proposed approach. If the community wants to drive a better solution in future we can always enhance the logic. On the other hand, if this is surprising to you and you expect to encounter issues with this as-is, then let's identify those issues and make sure this proposal is at least addressing your use case.

kvaster replied:
Think that idea with just allow or drop all vlan tagged packets with default to drop is better for now. It will defenitely solve problems with vlan we're having. Also I think that 'vlan aware' part of community isn't too big and this approach will suite everybody with such problem.

kvaster replied:
Also such approach will make much less changes to datapath.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -968,6 +976,21 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
+	if (ctx->vlan_present) {
+		vlan_id = ctx->vlan_tci & 0xfff;
+		if (vlan_id) {
+			if (allow_vlan(ctx->ifindex, vlan_id))
```
If I follow correctly, by default during upgrade, the new `--vlan-bpf-bypass` will be empty, which means that we will generate something like `#define VLAN_FILTER(x, y) { return false; }`. From your recent update on the main PR text, it sounds like this was the cause of your connectivity issue with your apiserver.

This sounds like the new default can break some connectivity in the network, which is not OK. That suggests that this entire new section needs to be conditionally run iff `--vlan-bpf-bypass` is configured.

joestringer replied:
Basically the "drop unconfigured VLANs" logic is OK if and only if there is no way for VLAN traffic to successfully reach its destination today. If some VLAN traffic may reach its destination but other VLAN traffic may end up malformed somewhere else on the network, then changing this code to drop VLAN-tagged traffic by default could cause existing setups to break. We don't want to change this behaviour and arbitrarily break some setups with the default configuration.

kvaster replied:
Sorry for my bad english...

`--vlan-bpf-bypass` will be empty by default, but empty does not mean 'block everything'. Empty allows vlan id's used by devices controlled by cilium. I've tried separately logic with just blocking everything and it's not good.

joestringer replied:
I worry that we are making this more complex than necessary for the default case to support this specific use case, and I'm already having trouble understanding the implications of the default behaviour with the PR right now.

The lowest-risk option here is to make no changes at all by default. So empty means do not try any special handling for the VLANs. Don't drop, don't bypass. Then separately, if you want "bypass for assigned devices" and/or "drop VLAN-tagged traffic coming in if it's not recognized or part of a list" then you can enable this shiny new option.

kvaster replied:
Just to clarify: current PR logic have no known for me issues.

joestringer replied:
@kvaster Are you configuring the new option? What about if you don't configure the new option?

kvaster replied:
@joestringer, my PR will always allow vlan id's used by devices controlled by cilium - by default traffic will be not blocked and will be controlled by cilium. You need to configure this option only if you have additional vlan devices which are not controlled by cilium (have bpf programs attached).

kvaster replied:
That means defaults with this PR (without configuring new option) should be enough for all reported cases in issue and should not break security.

joestringer replied:
> my PR will always allow vlan id's used by devices controlled by cilium - by default traffic will be not blocked and will be controlled by cilium. 

When you say "allow" here, do you mean `return CTX_ACT_OK`, ie "bypass Cilium logic"? That's the security concern I had in the previous issue that I was looking to avoid changing by default.

kvaster replied:
Yes, I mean `return CTX_ACT_OK`. And it will not bypass cilium logic, cause in this case packet will be processed by cilium by bpf program attached to vlan device - only vlan id's that will be processed by cilium are allowed by default... Generally we just don't care if we have additional devices and bpf programs are not attached to them. To make it proper we will have to rewrite vlan logic in bpf, but personally I think this is too much. Without such changes cilium will malform vlan tagged packets when attached to main device (to which vlan devices attached), cause it should not process such packets. And blocking every vlan tagged packet will lead to broken connectivity without possibility to investigate problem - probably this will happen only on first setup before users will know that they should explicitly allow vlan bpf bypass, but it will be hard to understand what is happening. I think that somebody from cilium team should decide what kind of behaviour is OK. I don't understand what can be done more to fix this issue properly. Thanks.

joestringer replied:
> And it will not bypass cilium logic, cause in this case packet will be processed by cilium by bpf program attached to vlan device 

I think you're basically saying, this logic is only bypassing Cilium BPF if it can identify that there is a VLAN device behind which will handle the traffic. I missed the nuance there initially, and at least in principle I agree that this mitigates the security concern we raised earlier since we have basically made a check to ensure that the traffic will be handled by Cilium after we skip processing.

My next question is, how do we know that the Cilium BPF program attached to the VLAN device is actually handling the traffic? Today the change is implemented through directly modifying `from_netdev()` / `to_netdev()` which is presumably used for both the BPF program on the main device and also on the VLAN device. Does this work because `ctx->vlan_present` is true only for VLAN packets arriving on the main device and it's false for decapsulated packets arriving on the vlan device?

kvaster replied:
Yes, it will work cause kernel will strip vlan tag and will process packet on vlan device - and bpf will process such packet cause vlan_present will be false already.

kvaster replied:
Also please note that allow_vlan function have ifindex argument. And if, for example, we will receive packet with double vlan tags then first one will be stripped, packet will be redirected for processing on vlan device and will be dropped there by allow_vlan function, cause no vlan tag is allowed on vlan device.

joestringer commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -245,3 +248,93 @@ func assertKeysInsideMap(c *C, m map[string]uint32, keys []string, want bool) {
+
+	option.Config.VLANBPFBypass = []int{4002, 4004, 4005}
+	m, err = vlanFilterMacros()
+	c.Logf("macros: %s", m)
```
It'd be good to turn this into a proper validation rather than a log message as well. `Logf()` doesn't do any testing and just makes the unit test output more noisy.

kvaster replied:
Oh, this log message was added my mistake, I will update PR. We just need to test for error here.

kvaster replied:
PR updated.

joestringer reviewed (approved):
I think this is good at this point, though I'd welcome additional review from others.

There's a couple of rough edges - No automatic adjustment for new VLAN devices showing up on the node, so you'd have to restart the Cilium agent. And if there are many VLAN devices on the node under a main VLAN device that's Cilium-managed then the user will be forced to configure this new option upon upgrade.. but I think that's probably within the bounds of reason. We can always continue to improve on this, I don't think these are grounds to hold up the whole fix.

kvaster committed:
Pass vlan tagged packets back to the kernel stack.

VLAN packets will be catched with bpf on main interface first. We
need to passs it back to the kernel stack. VLAN info/tag will be stripped and
packet will be reenqueued on proper interface or dropped.

Also we do not need to process VLAN packets on egress, cause either
we've already processed such packets in case bpf program is attached
to VLAN interface or we do not need to process such packets at all in
other way.

We will reject all unknown vlan tags for security reasons.
By default will be allowed only vlan tags of vlan devices controlled by
cilium. Additional vlan tags may be added using `vlan-bpf-bypass` option.
'0' tag (i.e. `--vlan-bpf-bypass 0`) can be used to allow all vlan tags without filtering.

Fixes: cilium#14579

Signed-off-by: Viktor Kuzmin <kvaster@gmail.com>

brb commented on bpf/bpf_host.c:
```diff
@@ -968,6 +976,21 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
+
+	/* Filter allowed vlan id's and pass them back to kernel.
+	 */
+	if (ctx->vlan_present) {
```
Have you checked whether this whole if-statement gets optimized away by the compiler if no vlans are specified? If not, then we would need to wrap this into #ifdef.

kvaster replied:
vlan_allow function is always inline and it defenitely will be optimized, cause the code will look like:

```
if (false) {
  return CTX_ACT_OK;
} else {
  return send_drop_notifiy(...);
}
```

But don't know how to check this...

brb replied:
You can use `llvm-objdump -S`. You can find the relevant object file it in `/var/run/cilium/state`. If it's not optimized, maybe the verifier will remove the branch. For this `bpftool prog dump` is useful.

kvaster replied:
Thanks. I will check this and will get back to you. Meanwhile checking all other changes for 'problems' will be appreciated :)

brb replied:
The rest LGTM. I'd appreciate if you would come up with a testing strategy too.

kvaster replied:
I'm testing locally on VM's, but my install is specific - gentoo linux... I think only testing on VM's will help here, but to write such test will need to learn a lot about cilium's testing infastructure...

kvaster replied:
I've checked - llvm-objdump is showing that this function is always optimized. If statements are removed in case they are not needed (always false or always true).

kvaster replied:
Marking this conversation as resolved cause of this:

```
       4:	54 01 00 00 ff 0f 00 00	w1 &= 4095
       5:	16 01 1b 00 00 00 00 00	if w1 == 0 goto +27 <LBB6_6>
       6:	b4 01 00 00 02 00 00 00	w1 = 2
       7:	63 16 40 00 00 00 00 00	*(u32 *)(r6 + 64) = r1
       8:	b4 01 00 00 4a ff ff ff	w1 = -182
       9:	63 16 38 00 00 00 00 00	*(u32 *)(r6 + 56) = r1
      10:	b4 01 00 00 00 00 00 00	w1 = 0
```

We can see that there is no ACT_OK return here. Function is optimized.

brb reviewed (approved):
Thanks!

nathanjsweet reviewed (approved):
From a hubble and api perspective. LGTM. The datapath stuff makes sense to me too.

errordeveloper reviewed (approved):
Protobuf and helm changes LGTM!

jrajahalme commented on pkg/datapath/linux/config/config.go:
```diff
@@ -549,6 +555,74 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+		// if it's vlan device and we're controlling vlan main device
+		// and either all vlans are allowed, or we're controlling vlan device or vlan is explicitly allowed
+		if ok && devices[vlan.ParentIndex] && (devices[vlan.Index] || allowedVlans[vlan.VlanId]) {
+			vlansByIfIndex[vlan.ParentIndex] = append(vlansByIfIndex[vlan.ParentIndex], vlan.VlanId)
```
This appending of vlanIDs that may not be ordered introduced a new unit test flake #17104.

joestringer replied:
@kvaster would you be able to look into this? Usually the solution is just to sort the slice once it's been assembled but before it is used to generate code.

joestringer replied:
Never mind, looks like @jrajahalme beat you to it :) Feel free to review though: https://github.com/cilium/cilium/pull/17105

kvaster replied:
Was away... Seems problem was already solved... I will make one more PR soon as discussed - docs update.
